Add support for Apple M2

enen92 opened the pull request:
Hey @narugit thanks for resurrecting the old smctemp project.
This PR adds support for Apple M2 (my laptop) (8 cores - 4 performance + 4 efficiency). Also adds support for other models based on the sensors app open-source code.
Closes #11

Also adds an option to poll for the temperature values for n times until a given value is returned (special case for some apple m2 models)

Mac Model:
<img width="278" alt="image" src="https://github.com/narugit/smctemp/assets/7375276/c9e95ec8-300c-4da0-b065-e58b27cc93cf">

TG Pro details:
<img width="799" alt="image" src="https://github.com/narugit/smctemp/assets/7375276/57cc5e61-ada8-438d-addf-b41c493f21b4">

smctemp:
<img width="286" alt="image" src="https://github.com/narugit/smctemp/assets/7375276/bc5f79cc-d5e8-48eb-8781-96d401b2520a">

enen92 committed:
Add support for Apple M2

enen92 commented on smctemp.cc:
```diff
@@ -388,35 +401,68 @@ double SmcTemp::GetCpuTemp() {
+    sensors.emplace_back(static_cast<std::string>(kSensorTp0j));
+    // CPU efficient core 4 temperature
+    sensors.emplace_back(static_cast<std::string>(kSensorTp0r));
+  } else { // Apple M1
```
This looks like the M1 pro or M1 max, probably not a bad idea to else if with the CPU model and just error out if the model is not supported

narugit replied:
> probably not a bad idea to else if with the CPU model and just error out if the model is not supported

It seems reasonable for me. Could you modify code like below?
```console
if (getCPUModel() == "Apple M1") { // Apple M1
  ...
} else if (getCPUModel() == "Apple M2") { // Apple M2
  ...
} else { // Not supported
  // return 0.0 to indicate that the CPU is not supported
  return temp;
}
```

enen92 replied:
Done in an equivalent way (to account for finer detection of the cpu model)

narugit commented:
@enen92 
Could you kindly explain how you obtained the sensor information in M2?
For example, how do you know `Tp0n` indicates CPU performance core 4 temperature？

joshuataylor commented:
Alright so `getCPUModel()` returns "Apple M2 Max" for me, willing to bet Ultra and other marketing names/product names such as Ultra etc will be "Apple M2 Ultra".

Also thinking the same for M1? But the else should catch it. Maybe in future some sort of case statement to check the first 8 characters or so?

Maybe it's best to check that it starts with "Apple M2"?

Not sure of how to do this properly in C++, but changing it to the following fixed it sometimes for me:
```
  // Apple M2 CPU Models begin with "Apple M2", then have the product name such as Max/Ultra after it
  if (getCPUModel().substr(0, 8) == "Apple M1") {
```

It'll return like "36.1" sometimes, then 4.2.. is it picking a value at random?

enen92 commented:
> Alright so `getCPUModel()` returns "Apple M2 Max" for me, willing to bet Ultra and other marketing names/product names such as Ultra etc will be "Apple M2 Ultra".
> 
> Also thinking the same for M1? But the else should catch it. Maybe in future some sort of case statement to check the first 8 characters or so?
> 
> Maybe it's best to check that it starts with "Apple M2"?
> 
> Not sure of how to do this properly in C++, but changing it to the following fixed it sometimes for me:
> 
> ```
>   // Apple M2 CPU Models begin with "Apple M2", then have the product name such as Max/Ultra after it
>   if (getCPUModel().substr(0, 8) == "Apple M2") {
> ```
> 
> It'll return like "36.1" sometimes, then 4.2.. is it picking a value at random?

I reworked it a bit on the last commit, it should now be detected. Note that those models have a lot more cpu cores (8 performance cores + 4 efficiency cores) so there are some sensors that are not picked up in the calculation.
I also added a commented block to add them later (I have no way to reverse engineer this/no hardware).
The average calculation should fix the "double" problem but a lot of sensors are missing for those models.

enen92 commented:
> @enen92 Could you kindly explain how you obtained the sensor information in M2? For example, how do you know `Tp0n` indicates CPU performance core 4 temperature？

I got this from the stats app: https://github.com/exelban/stats/blob/135eacec905a8d1d43259660f1c4a85a1838d376/Modules/Sensors/values.swift#L253-L262

Also checked against TG pro that the values are pretty similar so I guess the match is right

joshuataylor commented:
Awesome thanks, I'll check it and and double check the sensor values. I'll share here once I know more.

joshuataylor commented:
https://github.com/exelban/stats/issues/1286#issuecomment-1666938089

Also from my observations it looks really all over the place. TG Pro must have figured out which sensors to use... 😕

enen92 commented:
> [exelban/stats#1286 (comment)](https://github.com/exelban/stats/issues/1286#issuecomment-1666938089)
> 
> Also from my observations it looks really all over the place. TG Pro must have figured out which sensors to use... 😕

Anyway, thanks for the link. I removed those definitions from the comments to avoid confusion (I also don't know if they are an exact match). As this tool only computes the average temperature I'd say it's fine.

narugit commented on smctemp.cc:
```diff
@@ -33,6 +33,22 @@
+
+namespace {
+std::string getCPUModel() {
+  std::array<char, 512> buffer;
```
Could you add `#include <array>` above?

narugit commented:
@enen92 
Thank you for quick fix!
I've verified working well on Intel chip and M1 chip.

narugit commented:
@joshuataylor 
Thanks for trying!
Now that the feature branch is fixed, could you check if it works again on your Apple M2 Max?

joshuataylor commented:
Works randomly, but looks like these bytes randomly return the wrong value. Maybe merge this, then figure out the correct sensors for the Max?
```
  Tp05  [flt ]  4.6 (bytes: 89 41 94 40)
  Tp09  [flt ]  6.7 (bytes: 66 66 D6 40)
  Tp0D  [flt ]  4.6 (bytes: 89 41 94 40)
  Tp0f  [flt ]  6.7 (bytes: 66 66 D6 40)
  Tp0j  [flt ]  4.6 (bytes: 89 41 94 40)
```

joshuataylor commented:
[stats]https://github.com/exelban/stats) has the same issue, they disable the Max by default it seems.
<img width="292" alt="SCR-20230828-ttwo" src="https://github.com/narugit/smctemp/assets/225131/bb8bd4b9-6d46-46c5-865f-df6e505d6e6b">

TG Pro:
<img width="315" alt="SCR-20230828-ttxo" src="https://github.com/narugit/smctemp/assets/225131/97e80e4c-f1ff-48af-ba0d-21ca181e6090">

enen92 commented:
Yeah in sensors (last version) they seem to discard the values if lower than 10:

https://github.com/exelban/stats/commit/14e29c4d60229c363cca9c9d25c30c87b7870830#diff-ce3fe5eb966ca7c489e6b4db8d0d3a2441f0f8425779776f243611f9630cdafcR124

However the use-case for this app is different than smctemp. Sensors store state and display it to the user.
Smctemp pools the registers if requested by the user. If using this from the terminal (using watch or similar) maybe you should consider storing the last value or only display it if higher than the threshold.

enen92 committed:
Introduce valid limits for temperature sensors (arm)

enen92 committed:
Factor common code into CalculateAverageTemperature

enen92 committed:
Add option to poll the temperature for n times until a valid value is found

enen92 commented:
@narugit @joshuataylor ok so I've reworked this a bit and added a few more commits:

[510e806](https://github.com/narugit/smctemp/pull/14/commits/510e8061c9dbca85629670bea7e04ad611bc573d) -> Adds support for temperature limits to consider a valid sensor value valid range. To mimic [sensors](https://github.com/exelban/stats/commit/14e29c4d60229c363cca9c9d25c30c87b7870830#diff-ce3fe5eb966ca7c489e6b4db8d0d3a2441f0f8425779776f243611f9630cdafcR124) behaviour I've defined them as 10-120.

[b4b2e3b](https://github.com/narugit/smctemp/pull/14/commits/b4b2e3bafefb4ab895473237a4ef328734350e7f) -> Factors out common average temperature calculation in its own member function (avoid code duplication)

[b07a656](https://github.com/narugit/smctemp/pull/14/commits/b07a656a8a6f255cab0b4ef1ef0a556a705fbe1b) -> Extends the -c option to allow for arguments so that users can specify the number of request attempts. So, you can use for example `smctemp -c5` to retry for 5 times (1 second interval between attempts) to obtain the sensor values. As soon as a valid value is obtained, the result is returned (i.e. you might actually just do 1 request). That's the best way I've found to tackle the possible incorrect return values without forcing everyone to poll it ad-eternum.

What do you think?

joshuataylor commented:
The check time is awesome! Great idea!

It works adding a higher number, as I just use this for tmux I'll just check less often :-).

I was thinking before as well, maybe it would be good to add a flag for defining which sensors you want to check, but that's outside the scope of the PR and the sensors are currently hardcoded.

enen92 commented:
Thanks @joshuataylor, updated on the last revision.

joshuataylor commented:
Awesome! Yeah, that seems to have done the trick, always eventually getting a number back.

narugit commented:
@enen92 
Your idea of providing the option for polling is great, and I appreciate it.

I'd like to suggest using the option `-n` instead of `-c`. 
I am considering phasing out the `-c` option in the future. When running the `smctemp` command without any options, I plan to show Celsius temperature values in default.

enen92 commented:
> @enen92 Your idea of providing the option for polling is great, and I appreciate it.
> 
> I'd like to suggest using the option `-n` instead of `-c`. I am considering phasing out the `-c` option in the future. When running the `smctemp` command without any options, I plan to show Celsius temperature values in default.

What if in the future the tool supports more classes than just the CPU temperature? For example, the GPU? (which I have in the queue)

narugit commented:
How about something like this? (t is type of t)
- `-t [cpu,gpu,...]`: Show temperature of specified one (Default is CPU temperature)

narugit commented:
Ah,,

I just realized that you're suggesting whether it's possible to allow setting the polling frequency independently for each sensor when multiple sensors are specified.

narugit commented:
@enen92 
Even in the future, when the capability to specify multiple sensors becomes available, I believe there's no need to allow specifying the polling frequency for each sensor individually. It should suffice to have just one option for the polling frequency.

Since the polling frequency represents the maximum number of attempts, as long as we can retrieve a valid value within that specified number of attempts, there shouldn't be any issues.

For instance, if one wishes to configure the CPU temperature to poll 5 times and the GPU temperature 2 times, simply setting `-n5` should suffice. This assumes that the GPU can likely obtain the value within 2 attempts. I believe this approach can meet this user's requirements.

enen92 commented:
Hum, I think I misunderstood the request first (thought you meant to use `./smctemp -n5` to get the cpu temp with 5 attempts). In your last comment I think you actually meant `./smctemp -c -n5`. If that is the case, I agree - we can still define an operation (CPU) but an additional option for attempts (while not limiting future extensions).
If so, I'll try to have a crack at it tomorrow, already too late here :)

narugit commented:
@enen92 
Actually, `-c` doesn't refer to CPU; it stands for Celsius temperature. I apologize for any confusion caused by having both CPU and Celsius in the same sentence.
https://github.com/narugit/smctemp/blob/da1fe798b9ebe209be820b557815e37c6152a0da/README.md#L23

And as you've correctly stated, the expectation behavior after the merging of this PR is to use the command `smctemp -c -n5` to get the cpu temp (in Celsius) with 5 attempts.

enen92 commented:
> @enen92 Actually, `-c` doesn't refer to CPU; it stands for Celsius temperature. I apologize for any confusion caused by having both "C"PU and "C"elsius in the same sentence.
> 
> https://github.com/narugit/smctemp/blob/da1fe798b9ebe209be820b557815e37c6152a0da/README.md#L23
> 
> And as you've correctly stated, the expectation behavior after the merging of this PR is to use the command `smctemp -c -n5` to get the cpu temp (in Celsius) with 5 attempts.

Should be done in the last push.
Actually -c in the code sets the operation to read the CPU values so I thought it was dedicated for the CPU. If introducing support for other units maybe it's not a bad idea to introduce (-C uppercase or -F) in the future.

narugit commented:
@enen92 
> If introducing support for other units maybe it's not a bad idea to introduce (-C uppercase or -F) in the future.

Thank you for the great idea. Sounds great!

narugit commented on main.cc:
```diff
@@ -60,7 +73,17 @@ int main(int argc, char *argv[]) {
+        temp = smc_temp.GetCpuTemp();
+        attempts--;
+        if (attempts > 0 && (temp == 0.0 || std::isnan(temp))) {
+          usleep(1000000);
```
Let's add single quote to be easier to read.
```suggestion
          usleep(1'000'000);
```

enen92 replied:
Personal preference but added it anyway

narugit commented on smctemp.cc:
```diff
@@ -366,6 +382,22 @@ kern_return_t SmcAccessor::PrintAll() {
+      valid_sensor_count++;
+    }
+  }
+  temp /= valid_sensor_count;
```
If all sensor values are out of limits, zero-devided will happen.
```suggestion
  if (valid_sensor_count > 0) {
    temp /= valid_sensor_count;
  }
```

narugit commented on main.cc:
```diff
@@ -60,7 +73,17 @@ int main(int argc, char *argv[]) {
     case smctemp::kOpReadCpuTemp:
-      std::cout << std::fixed << std::setprecision(1) << smc_temp.GetCpuTemp();
+      double temp = 0.0;
+      while (attempts > 0) {
+        temp = smc_temp.GetCpuTemp();
+        attempts--;
+        if (attempts > 0 && (temp == 0.0 || std::isnan(temp))) {
+          usleep(1000000);
+        } else {
+          break;
+        }
+      }
```
It seems like we can simplify the logic. How about this?
```suggestion
      while (attempts > 0) {
        temp = smc_temp.GetCpuTemp();
        if (temp > 0.0) {
          break;
        } else {
          usleep(1000000);
          attempts--;
        }
      }
```

narugit commented on main.cc:
```diff
@@ -11,22 +12,33 @@ void usage(char* prog) {
   smctemp::SmcVal_t      val;
 
-  while ((c = getopt(argc, argv, "clvh")) != -1) {
+  while ((c = getopt(argc, argv, "clvhn::")) != -1) {
```
It seems that double colons are a GNU extension, and on one of my two Macs unfortunately, this doesn't work as expected. 
```console
$ ./smctemp -n1 -c
65.8

$ ./smctemp -n 1 -c
Check Temperature by using Apple System Management Control (Smc) tool 0.1.1
Usage:
./smctemp [options]
    -c         : list CPU temperatures (Celsius)
    -h         : help
    -l         : list all keys and values
    -v         : version
    -n         : tries to query the temperature sensor for n times (1 second interval) until a valid value is returned
```

Ref. https://man7.org/linux/man-pages/man3/getopt.3.html#DESCRIPTION
> Two colons
>       mean an option takes an optional arg; if there is text in the
>       current argv-element (i.e., in the same word as the option name
>       itself, for example, "-oarg"), then it is returned in optarg,
>       otherwise optarg is set to zero.  This is a GNU extension. 

How do you think about using a single colon and making the argument for -n mandatory?
```suggestion
  while ((c = getopt(argc, argv, "clvhn:")) != -1) {
```

enen92 replied:
I think it is fine. The only drawback is that it somehow breaks the "contract" for those already using simply `./smctemp -c`.
But I personally think it's fine since the project is still quite new. I'll adjust it in a few hours.

narugit commented:
Thank you for your quick modification!
I've checked the behavior on my Intel and M1 mac and it works:)
It is the same behavior as before on both Intel and M1 platforms with `-c`, while also allowing the use of the `-n` option.

narugit commented:
@joshuataylor 
Could you please check the behavior is as you expected on your M2 Max?

joshuataylor commented:
Apologies for the delay, I missed the @ and the follow up here! 🤦 
Everything seems to be good on my end! 🎉

narugit commented:
@joshuataylor 
Thank you for confirming! 

@enen92 
Thank you for a great PR 🎉

joshuataylor commented:
Sorry for the necro, but also looks like Apple may have FINALLY fixed the temperature reporting for M2's, as of Ventura 13.6.2 it seems to work really consistently now!
